pkg/rbac/userData.go

```diff
@@ -18,41 +18,113 @@ import (
 // Contains data about the resources the user is allowed to access.
 type userData struct {
 	// clusters     []string              // Managed clusters where the user has view access.
-	// csResources  []resource            // Cluster-scoped resources on hub the user has list access.
+	csResources []resource            // Cluster-scoped resources on hub the user has list access.
 	nsResources map[string][]resource // Namespaced resources on hub the user has list access.
 
-	// Internal fields to manage the cache.
-	// clustersErr       error      // Error while updating clusters data.
-	// clustersLock      sync.Mutex // Locks when clusters data is being updated.
-	// clustersUpdatedAt time.Time  // Time clusters was last updated.
-	err     error      // Error while getting user data from cache
-	csrErr  error      // Error while updating cluster-scoped resources data.
-	csrLock sync.Mutex // Locks when cluster-scoped resources data is being updated.
-	// csrUpdatedAt time.Time  // Time cluster-scoped resources was last updated.
+	csrErr       error      // Error while updating cluster-scoped resources data.
+	csrLock      sync.Mutex // Locks when cluster-scoped resources data is being updated.
+	csrUpdatedAt time.Time  // Time cluster-scoped resources was last updated.
 	nsrErr       error      // Error while updating namespaced resources data.
 	nsrLock      sync.Mutex // Locks when namespaced resources data is being updated.
 	nsrUpdatedAt time.Time  // Time namespaced resources was last updated.
-
-	// impersonate *kubernetes.Interface // client with impersonation config
+	authzClient  v1.AuthorizationV1Interface
 }
 
-func (cache *Cache) GetUserData(ctx context.Context, clientToken string) (*userData, error) {
-	var user userData
+func (cache *Cache) GetUserData(ctx context.Context, clientToken string, authzClient v1.AuthorizationV1Interface) (*userData, error) {
+	var user *userData
 	uid := cache.tokenReviews[clientToken].tokenReview.Status.User.UID //get uid from tokenreview
 	cache.usersLock.Lock()
 	defer cache.usersLock.Unlock()
 	cachedUserData, userDataExists := cache.users[uid] //check if userData cache for user already exists
-	if userDataExists {
+	// UserDataExists and its valid
+	if userDataExists && userCacheValid(cachedUserData) {
 		klog.V(5).Info("Using user data from cache.")
-		return cachedUserData, user.err
-
+		return cachedUserData, nil
+	} else {
+		// User not in cache , Initialize and assign to the UID
+		user = &userData{}
+		cache.users[uid] = user
+		// We want to setup the client if passed, this is only for unit tests
+		if authzClient != nil {
+			user.authzClient = authzClient
+		}
 	}
-
 	userData, err := user.getNamespacedResources(cache, ctx, clientToken)
+
+	// Get cluster scoped resources for the user
+	// TO DO : Make this parallel operation
+	if err == nil {
+		klog.V(5).Info("No errors on namespacedresources not present ....Computing now", user)
```

Comment:
```suggestion
		klog.V(5).Info("No errors on namespacedresources present ....Computing now", user)
```

Reply:
Also, do we need to log the user info? Maybe we can but at a higher verbosity?
```suggestion
		klog.V(6).Info("No errors on namespacedresources present ....Computing now", user)
```

Reply:
If we want to log that , as well see the user too right ? Is this high priority - or can hold @Anxhela21